### DISCUSSION TITLE
how to reproduce the exclusion constraint violation error and possible solutions

### DISCUSSION THREAD

**Discussion by manthan-jsharma** [mathesar-foundation/mathesar#3193] [Developer Q&A]
Hi, 

I'm currently working on #2126 

i have opened the mathesar project in demo mode, but unaware where to run the sql query or is there any other way to implement it using the frontend @mathemancer ?

```
CREATE EXTENSION IF NOT EXISTS btree_gist;
ALTER TABLE "Checkouts" DROP CONSTRAINT IF EXISTS sim_chck_excl;
ALTER TABLE "Checkouts"
  ADD CONSTRAINT sim_chck_excl
  EXCLUDE USING gist
    ("Item" WITH =, tsrange("Checkout Time", "Check In Time", '[]') WITH &&);
```

cc  @dmos62

**Comment by dmos62** [comment]
To reproduce, you'll want to log into demo mode via the frontend, so that you have a demo session. Then connect to the demo database via pgadmin or psql and use that query to alter the library schema. Then try to use that schema via the frontend.

**Comment by manthan-jsharma** [comment]
hey @dmos62,

I'm able to reproduce the constraint violation error in `"checkout"` table.

<img width="628" alt="Screenshot 2023-09-05 at 2 03 00 PM" src="https://github.com/centerofci/mathesar/assets/123374300/99b34cc2-401b-487b-a296-a0d50e7acc41">'

Yes I agree, the error message is not user friendly, Below are some potential error messages that we can consider displaying:
\
&nbsp;

1. The first solution is to display the same error message with **enhanced formatting**, achieved by incorporating appropriate spaces.

<img width="696" alt="Screenshot 2023-09-05 at 3 07 11 PM" src="https://github.com/centerofci/mathesar/assets/123374300/c86d7c55-35d6-42ec-8ee0-a3e170868339">

\
&nbsp;

2. The second solution involves incorporating a **customised error message** specifically for cases when an exclusion constraint is triggered.

<img width="612" alt="Screenshot 2023-09-05 at 3 10 38 PM" src="https://github.com/centerofci/mathesar/assets/123374300/82aec0cd-2bd7-4874-9413-fac184ed5a4d">

\
&nbsp;
I believe that the 2nd solution is quite challenging to implement and may lack generality. In my opinion, I would prefer to implement the 1st solution, which involves displaying the error message with improved formatting and line breaks.

@dmos62 , @mathemancer , please let me if you have any other solutions to discuss and and also please confirm whether we should implement first solution or not ?

**Reply by dmos62** [reply]
Examine how we're currently handling violations of the unique constraint, for example. That will be the best starting point.

**Comment by manthan-jsharma** [comment]
hi @dmos62,

I have raised a pull request. Please review it: https://github.com/centerofci/mathesar/pull/3200


<img width="1221" alt="Screenshot 2023-09-08 at 11 48 10 PM" src="https://github.com/centerofci/mathesar/assets/123374300/fca3dc2a-c8fa-4f8b-a354-9dff83505fb4">

\
&nbsp;

can you please confirm if we need any extra data in detail property in api response ? 

<img width="724" alt="Screenshot 2023-09-09 at 12 02 40 AM" src="https://github.com/centerofci/mathesar/assets/123374300/b0b2dc05-2742-4303-b8ea-ca196abc4941">

\
&nbsp;

**I will write the test cases once the changes are confirmed**
